web_autoflip_demo/autoflip_demo_UI/src/download.ts

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright 2020 Google LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  cropWindowStorage,
+  videoBuffer,
+  audio,
+  updateAudio,
+  outputStorage,
+} from './globals';
+
+import {
+  videoPreview,
+  canvas,
+  videoRecord,
+  card2,
+  card4,
+  timerDisplay,
+} from './globals_dom';
+
+import { ffmpegWorkerCombine, ffmpegWorkerAudio } from './globals_worker';
+export function createDownload(inputWidth: number, inputHeight: number): void {
+  // Puases the main preview video player to ensure recording quality
+  videoPreview.pause();
+  const canvas2D = canvas.getContext('2d') as CanvasRenderingContext2D;
+
+  const cropWindows: ExternalRenderingInformation[] =
+    cropWindowStorage[`${inputHeight}&${inputWidth}`];
+  const render = cropWindows[0].renderToLocation as Rect;
+  canvas.width = render.width;
+  canvas.height = render.height;
+
+  let videoCropX = 0;
+  let videoCropY = 0;
+  let videoCropWidth = 1920;
+  let videoCropHeight = 1080;
+  let mediaRecorder: MediaRecorder;
+  let recordedBlobs: Blob[];
+  let output: ArrayBuffer;
+  let stream: any;
+
+  videoPreview.currentTime = 0;
+  videoRecord.play();
+  startRecording();
+  loop();
+
+  function loop() {
+    canvas2D.imageSmoothingEnabled = false;
+    if (!videoRecord.paused && !videoRecord.ended) {
+      canvas2D.drawImage(
+        videoRecord,
+        videoCropX,
+        videoCropY,
+        videoCropWidth,
+        videoCropHeight,
+        0,
+        0,
+        render.width,
+        render.height,
+      );
+      stream.getVideoTracks()[0].requestFrame();
+      setTimeout(loop, 35);
+    } else {
+      stopRecording();
+    }
+  }
+
+  function handleDataAvailable(event: BlobEvent): void {
```

Comment:
add small js docs to the functions. (js function comments)

Reply:
fixed